packages/amesos2/src/Amesos2_TpetraMultiVecAdapter_def.hpp

```diff
@@ -143,8 +144,53 @@ namespace Amesos2 {
     // Redistribute the input (multi)vector.
     redist_mv.doExport (*mv_, *exporter_, Tpetra::REPLACE);
 
+    if ( distribution != NON_CONTIGUOUS_AND_ROOTED ) {
+    // Do this if GIDs contiguous - existing functionality
     // Copy the imported (multi)vector's data into the ArrayView.
-    redist_mv.get1dCopy (av, lda);
+      redist_mv.get1dCopy (av, lda);
+    }
+    else {
+    // Do this if GIDs not contiguous...
+      // sync is needed for example if mv was updated on device, but will be passed through Amesos2 to solver running on host
+      typedef typename multivec_t::dual_view_type dual_view_type;
+      typedef typename dual_view_type::host_mirror_space host_execution_space;
+      redist_mv.template sync < host_execution_space > ();
+
+      auto contig_local_view_2d = redist_mv.template getLocalView<host_execution_space>();
+      if ( redist_mv.isConstantStride() ) {
+        for ( size_t j = 0; j < num_vecs; ++j) {
+          auto av_j = av(lda*j, lda);
+          for ( size_t i = 0; i < lda; ++i ) {
+            av_j[i] = contig_local_view_2d(i,j); //lda may not be correct if redist_mv is not constant stride...
+          }
+        }
+      }
+      else {
+        // ... lda should come from Teuchos::Array* allocation,
+        // not the MultiVector, since the MultiVector does NOT
+        // have constant stride in this case.
+        // TODO lda comes from X->getGlobalLength() in solve_impl - should this be changed???
+        const size_t lclNumRows = redist_mv.getLocalLength();
+        for (size_t j = 0; j < redist_mv.getNumVectors(); ++j) {
+          auto av_j = av(lda*j, lclNumRows);
+          auto X_j = redist_mv.getVector(j);
+          auto X_lcl_j_2d = redist_mv.template getLocalView<host_execution_space> ();
+          auto X_lcl_j_1d = Kokkos::subview (X_lcl_j_2d, Kokkos::ALL (), j);
+          for ( size_t i = 0; i < lclNumRows; ++i ) {
```

Comment:
@mhoemmen thanks for the suggested change for the non-constant stride case when copying from the MultiVector to the Teuchos::Array. 
If you get a chance to take a glance and notice anything I missed please let me know.